### ISSUE TITLE
Optimization for scheduler preemption

### ISSUE THREAD

**Issue by hex108** [kubernetes/kubernetes#89036]
**What would you like to be added**:

* If a pod's priority is lowest(e.g. there is three priroites in cluster: 10, 20 and 30, and the pod's priority is 10), it could not preempt any other pods, then we could skip trying preemption for it, it will improve schedule performance. 

* We could just try to search a part of nodes instead of all nodes for preemption, the idea is inspired by [Percentage of Nodes to Score
](https://kubernetes.io/docs/concepts/scheduling/scheduler-perf-tuning/#percentage-of-nodes-to-score)

**Why is this needed**:

In our clusters(>3K nodes), preemption occurs frequently, scheduler could only try to schedule less than 10 pods. It could be improved by above ideas.

/sig scheduling
/assign

**Comment by Huang-Wei** [comment]
> If a pod's priority is lowest(e.g. there is three priroites in cluster: 10, 20 and 30, and the pod's priority is 10), it could not preempt any other pods, then we could skip trying preemption for it, it will improve schedule performance.

I can see the point to skip preemption for Pods with default priority (0). However, in your case, Pods with priority 10 can still preempt Pods such as the ones with default priority, how can you manage to skip them?

> We could just try to search a part of nodes instead of all nodes for preemption

This is a valid point. In the next release (1.19), we will extract the hard-coded preemption logic into plugins, which is a good timing to consider introducing some pluginConfig argument like `PercentageOfNodesToTryPreempt`.

**Comment by hex108** [comment]
> I can see the point to skip preemption for Pods with default priority (0). However, in your case, Pods with priority 10 can still preempt Pods such as the ones with default priority, how can you manage to skip them?

The priroties(10, 20, and 30) are just examples. Actually, the priorities could be any interger number and we can set the default priority to any valid value by [setting one priority class's globalDefault value to true](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass). We could list all prioriy classes and find out which is the lowest priroity class.(For optimizaiton, we could record the lowest priority using a variable, and create a prirority informer to wathc priority class add/update/delete event, and update the vairable accordingly.)

> This is a valid point. In the next release (1.19), we will extract the hard-coded preemption logic into plugins, which is a good timing to consider introducing some pluginConfig argument like PercentageOfNodesToTryPreempt.

Glad to know it! Since they are not dependent, maybe we could start the work for `PercentageOfNodesToTryPreempt` in advance?

**Comment by Huang-Wei** [comment]
> The priroties(10, 20, and 30) are just examples. Actually, the priorities could be any interger number and we can set the default priority to any valid value by [setting one priority class's globalDefault value to true](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass). We could list all prioriy classes and find out which is the lowest priroity class.(For optimizaiton, we could record the lowest priority using a variable, and create a prirority informer to wathc priority class add/update/delete event, and update the vairable accordingly.)

I see the point.

> maybe we could start the work for PercentageOfNodesToTryPreempt in advance?

As you know, current preemption logic is hard-coded, to avoid duplicate/interim efforts, I'd like to come up with a general write-up on preemption extension point first, within which let's discuss propose possible args and other aspects.

**Comment by hex108** [comment]
> As you know, current preemption logic is hard-coded, to avoid duplicate/interim efforts, I'd like to come up with a general write-up on preemption extension point first, within which let's discuss propose possible args and other aspects.

Get it now. Thanks!

**Comment by alculquicondor** [comment]
Are you sure this is the actual bottleneck? I could imagine some Filter plugins being already as slow as what the preemption logic does already: linearly check all pods priorities and stop if it finds none with lower priority (`selectVictimsOnNode`).

1. Could you try to profile this?
1. I don't think we should have this as a plugin, but rather in the core scheduler.

**Comment by Huang-Wei** [comment]
> I don't think we should have this as a plugin, but rather in the core scheduler.

Jun raised 2 points here, @alculquicondor could you clarify which one (or both) you don't think should be exposed as plugin?

**Comment by hex108** [comment]
> Are you sure this is the actual bottleneck

The logs show that preemption takes most of the time. I could do some profile for it . Howerver there is another problem that I try to solve: `If you delete a PriorityClass, existing Pods that use the name of the deleted PriorityClass remain unchanged`([reference here](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass)). It means that it is not enough to just list priority classes to find the lowest priority class.

> If you delete a PriorityClass, existing Pods that use the name of the deleted PriorityClass remain unchanged

Which filter plugin? Maybe we could do some optimization for it too :)

**Comment by Huang-Wei** [comment]
> If you delete a PriorityClass, existing Pods that use the name of the deleted PriorityClass remain unchanged

This may be implemented in garbage collector when it detects a PriorityClass gets deleted.

**Comment by hex108** [comment]
> This may be implemented in garbage collector when it detects a PriorityClass gets deleted.

Do you mean that garbage collector deletes pods that use the deleted priroirty class?

**Comment by Huang-Wei** [comment]
Yes, which is running as a watchdog goroutine in controller manager.

**Comment by hex108** [comment]
> Yes, which is running as a watchdog goroutine in controller manager.

I am not sure it makes sense because users does not know whether there are pods using the priroity class when they delete it, and the deleted class acutually should not affect existing pods.

How about counting and recording the number of pods that use each priroity class?  Then we could find the lowest priority.

**Comment by alculquicondor** [comment]
> The logs show that preemption takes most of the time.

I didn't mean that preemption might not be the bottleneck. It likely is. However,

> if a pod's priority is lowest it could not preempt any other pods

I don't think that's the scenario that is going to save us much running time. If there is at least one pod per node that has lowest priority, we still have to do the rest of the preemption algorithm.

> We could just try to search a part of nodes instead of all nodes for preemption, the idea is inspired by Percentage of Nodes to Score

I like this (and it could be part of the core scheduler, rather than in a plugin). If we find enough nodes were preemption works, we could stop. However, if preemption doesn't help anywhere, we would still check all nodes (but maybe it would stop fast on some).

I suggest we build a representative perf test case, do profiling and find the bottlenecks.

**Comment by Huang-Wei** [comment]
> I am not sure it makes sense because users does not know whether there are pods using the priroity class when they delete it, and the deleted class acutually should not affect existing pods.

It's debatable. I read the priorityClass <-> pod.spec.priority as a "soft" controllerRef - priorityClass acting as the controller, pod.spec.priority as the controllee. Once the controller gets updated/deleted, the controllee should be impacted in a cascading way.

**Comment by ahg-g** [comment]
> I don't think that's the scenario that is going to save us much running time. If there is at least one pod per node that has lowest priority, we still have to do the rest of the preemption algorithm.

That is not true, pods in the same priority class don't evict each other. So a pod with the lowest priority will never benefit from running the preemption logic.

> If a pod's priority is lowest(e.g. there is three priroites in cluster: 10, 20 and 30, and the pod's priority is 10), it could not preempt any other pods, then we could skip trying preemption for it, it will improve schedule performance.

Ideally, the scheduler will need a heap to track the lowest priority in the system, and for each priority we keep track of the number of pods so that we can remove it from the heap when the count reduces to zero. 

However, one solution could be to set ```pod.Spec.PreemptionPolicy=PreemptNever``` for pods with the lowest priority class, a mutating webhook should do the trick if you know the lowest priority class in the system beforehand.

> We could just try to search a part of nodes instead of all nodes for preemption, the idea is inspired by Percentage of Nodes to Score

+1 to limiting the number of nodes examined during preemption.

**Comment by Huang-Wei** [comment]
> That is not true, pods in the same priority class don't evict each other. So a pod with the lowest priority will never benefit from running the preemption logic.

I suppose @hex108 meant that if a Pod can be determined to hold a (global) lowest priority, it even doesn't need to be compared with other available Pods across nodes.

**Comment by ahg-g** [comment]
> > That is not true, pods in the same priority class don't evict each other. So a pod with the lowest priority will never benefit from running the preemption logic.
> 
> I suppose @hex108 meant that if a Pod can be determined to hold a (global) lowest priority, it even doesn't need to be compared with other available Pods across nodes.

I agree, my comment was a reply to Aldo's comment which seemed to disagree with the statement "if a pod's priority is lowest it could not preempt any other pods"

**Comment by alculquicondor** [comment]
> I agree, my comment was a reply to Aldo's comment which seemed to disagree with the statement "if a pod's priority is lowest it could not preempt any other pods"

I don't disagree. I'm just saying that in that scenario preemption ends fast anyways. So we have to find the real bottlenecks.

**Comment by ahg-g** [comment]
> > I agree, my comment was a reply to Aldo's comment which seemed to disagree with the statement "if a pod's priority is lowest it could not preempt any other pods"
> 
> I don't disagree. I'm just saying that in that scenario preemption ends fast anyways. So we have to find the real bottlenecks.

agree with benchmarking, but the logic still iterates over all pods in the system and compares the priorities, and this is happening in the scheduling cycle, so in a large scale cluster with > 100k pods, this may be a problem.

**Comment by hex108** [comment]
> It's debatable. I read the priorityClass <-> pod.spec.priority as a "soft" controllerRef - priorityClass acting as the controller, pod.spec.priority as the controllee. Once the controller gets updated/deleted, the controllee should be impacted in a cascading way.

@Huang-Wei  It will be a incompatible change. And we need to check whether there are pods using the the pirority class and give a warning when users try to delete the priority class.  However it will make things simpler if it is acceptable.

>  Ideally, the scheduler will need a heap to track the lowest priority in the system, and for each priority we keep track of the number of pods so that we can remove it from the heap when the count reduces to zero.

@ahg-g Yes, I'd like to try this way. WDYT? @Huang-Wei 

> However, one solution could be to set pod.Spec.PreemptionPolicy=PreemptNever for pods with the lowest priority class, a mutating webhook should do the trick if you know the lowest priority class in the system beforehand.

It could be a solution if prirotiy classes are static(no priroty classes are deleted or added), because we need to update the `pod.Spec.PreemptionPolicy` of those pods using prior lowest priority class once a much lower priority class is added.

**Comment by alculquicondor** [comment]
@hex108 would you mind adding a perf test case in `test/integration/scheduler_perf/config/performance-config.yaml`?

**Comment by ahg-g** [comment]
>@ahg-g Yes, I'd like to try this way. WDYT? @Huang-Wei

As @alculquicondor suggested, we need a benchmark first to understand how big of a problem this really is.

**Comment by alculquicondor** [comment]
/remove-lifecycle stale

cc @adtac

**Comment by adtac** [comment]
tl;dr: I've modified defaultpreemption to only look at a moving window of all nodes to consider for preemption. While this is very much preliminary work (i.e. acceptability of the  nodes getting preempted needs to be verified), the results from a performance perspective are very promising. Commit can be found here: https://github.com/adtac/kubernetes/commit/  5906bcc0917c0fd7207d669163124e2eaace5b78

So I've been running the preemption integration benchmarks we have with profiling turned on and I noticed the following. All time values are percentages of the total test run time  (wall clock) of the 5k node test.

1. CPU spent on plugin accounts for 23% (about 7% in the main scheduler thread, 16% in parallel processing).
1. Of the 7% in the main thread, 5.6% is SelectCandidate, which picks one node from a pool of candidates.
1. Of the 16% in parallel processing, all of it is dry running preemption on all nodes, with each node's processing spending most of its time in selectVictimsOnNode.
1. Naturally, the 16% is a lower hanging fruit to optimise. So I modified the plugin to only dry run preemption on a moving window of nodes.
   a. Moving window size is 10% of the cluster size.
   b. The window size is always at least 100 nodes.
   c. The window accounts for resizes in cluster size.
   d. The window wraps around gracefully if the cluster size isn't a multiple of the window size.
   e. 10% window size and 100 minimum are preliminary heurestics. Feedback on better values would be greatly appreciated.
1. Impact of optimisation on throughput: 11.5 pods/sec before, 14.0 pods/sec after -- 21.7% improvement
1. Time spent on plugin code as a percentage of test run time: 23% before (7% main, 16% parallel), 6% after (2% main, 4% parallel) -- 74% lower
1. Unfortunately, the integration tests don't report the `plugin_execution_duration_seconds` metric yet.                                                                             

Will update with more details.

**Comment by alculquicondor** [comment]
Just to confirm, is the benchmark exercising a successful or unsuccessful preemption (meaning that no preemption of pods is possible)?

If the former, maybe we could use the same "percentageOfNodesToScore" approach (stop looking once a percentage of nodes would be schedulable after preempting some pods). But this wouldn't help the worst case scenario (unsuccessful preemption).

Other than that, before doing architectural changes, sometimes there is low hanging fruit that can be optimized (like removing locks, map lookups, parsing, etc.). Also, somewhat related to the initial proposal in this issue: look for skipping checking some pods. In particular, we don't sort the pods in a NodeInfo. Keeping them sorted by priority or some other criteria might help. Can you check for those?

**Comment by adtac** [comment]
>Just to confirm, is the benchmark exercising a successful or unsuccessful preemption (meaning that no preemption of pods is possible)?

The benchmark requires all 5000 pods to be successfully preempted to be scheduled. After scheduling 4 pods/node, each node has insufficient resources to schedule a measurement pod without preemption, so the fact that the test completes is proof that each measurement pod preempted one or more initial pod.

>If the former, maybe we could use the same "percentageOfNodesToScore" approach (stop looking once a percentage of nodes would be schedulable after preempting some pods). But this wouldn't help the worst case scenario (unsuccessful preemption).

Indeed. That approach should be similar to the approach I've taken with this experiment from a performance perspective, but should be a bit more robust in scenarios where preemption isn't possible.

>removing locks

I do have some ideas that I want to test (see TODO in the commit patch for example). But from what I can tell from preliminary experiments and reading the code, locks are sparsely used. There is some scope for making the whole approach lock-free though.

>look for skipping checking some pods

We do some skips (e.g., higher priority pods aren't checked, naturally), but perhaps more could be done there, ack.

>we don't sort the pods in a NodeInfo

We do sort potential victims by priority before parallel processing:

https://github.com/kubernetes/kubernetes/blob/b14fcc544bd5a1e6f595ebebf54754ffae3de87e/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L559

**Comment by alculquicondor** [comment]
> Indeed. That approach should be similar to the approach I've taken with this experiment from a performance perspective, but should be a bit more robust in scenarios where preemption isn't possible.

I don't think we should stop looking for nodes in the case were preemption isn't possible. There could be just one node that can be preempted and we want to do it, because the incoming pod has higher priority.

> We do sort potential victims by priority before parallel processing:

How expensive is that line? It might help to keep the pods sorted. But probably most of the cost is in running the filter plugins anyways :/

**Comment by adtac** [comment]
>I don't think we should stop looking for nodes in the case were preemption isn't possible. There could be just one node that can be preempted and we want to do it, because the incoming pod has higher priority.

Yes, I agree. I meant to say that with my current approach, preemption might not be possible when looking at only 10% of the nodes, but with the percentageOfNodesToScore approach you suggested, your suggestion will be more robust because it will exhaust all nodes before confirming that preemption isn't possible. However, in scenarios where preemption is possible (this benchmark, for example), there shouldn't be a performance difference between the two approaches. That's what I meant to say :)

>How expensive is that line? It might help to keep the pods sorted. But probably most of the cost is in running the filter plugins anyways :/

According to the flamegraphs, less than 0.1% of the runtime is spent on sorting the potential victims. To be fair, though, in this benchmark, there are only four low priority pods per node, but even nodes with more pods shouldn't spend much time on that.

**Comment by ahg-g** [comment]
My bigger concern here is the worst case scenario: executing preemption when it really can't help. Typically the number of priority classes in a cluster is just a handful, we should capitalize on that. One idea is to track for each priority the nodes with at least one pod in that priority.

-1 -> node-a, node-b
0 -> node-a, node-c
1000 -> node-c, node-b

A pod with a specific priority would iterate over the nodes with pods with a strictly higher priority. We can also apply the windowing approach over that.

with regards to sorting, we are still iterating overall pods few lines above: https://github.com/kubernetes/kubernetes/blob/b14fcc544bd5a1e6f595ebebf54754ffae3de87e/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L530

**Comment by adtac** [comment]
I've opened a PR to enumerate fewer candidates (improves the average case). As discussed offline and as mentioned by @ahg-g, the improvements made by this PR will be orthogonal to the one proposed in https://github.com/kubernetes/kubernetes/issues/89036#issuecomment-692817149 (which I'll be doing too, but later).

**Comment by hex108** [comment]
/remove-lifecycle stale

I'll try to work on case 1.

**Comment by alculquicondor** [comment]
@hex108 can you update the issue description to mark the things that were already done? or open a new issue for the pending task.

**Comment by adtac** [comment]
>My bigger concern here is the worst case scenario: executing preemption when it really can't help. Typically the number of priority classes in a cluster is just a handful, we should capitalize on that. One idea is to track for each priority the nodes with at least one pod in that priority.

I'm working on this. I have a couple of ideas I want to prototype and experiment with (e.g. a memoized cache to keep track of the lowest priority in each node that expires after N seconds).

**Comment by adtac** [comment]
/unassign
I don't have the time to work on this

**Comment by ahg-g** [comment]
/remote-lifecycle stale

I think it is still worth pursuing the idea of tracking the lowest pod priority in the cluster and short circuit preemption if the unschedulable pod belongs to that priority.

**Comment by ahg-g** [comment]
/help

I think it is worth doing https://github.com/kubernetes/kubernetes/issues/89036#issuecomment-692817149
